File: modularmotifs/dsl/_syntax/basic_ast.py
import abc

# TODO: I think we would want to replace TypeGuard with TypeIs if we were all using Python 3.13...
from typing import Union, TypeGuard, Self, Optional, Type, Any, List
from types import ModuleType
from modularmotifs.core import Motif, Design, Color, CompositeMotif
from modularmotifs.core.design import PlacedMotif
import modularmotifs.motiflibrary.examples as libexamples

# from pathlib import Path


class FreshVar:
    """A source of fresh variable names"""

    def __init__(self, base_name="x", names_to_avoid: Optional[set[str]] = None):
        """Initialize a new FreshVar instance

        Args:
            base_name (str, optional): the prefix for variable names. Defaults to "x".
            names_to_avoid (Optional[set[str]], optional): a set of variable names to avoid (i.e., because it would result in overwriting a previously defined variable). Defaults to None.
        """
        self.current = 0
        self.base_name = base_name
        self.names_to_avoid = names_to_avoid
        pass

    def get_fresh(self) -> str:
        fresh = f"{self.base_name}{self.current}"
        self.current += 1
        if self.names_to_avoid:
            while fresh in self.names_to_avoid:
                fresh = f"{self.base_name}{self.current}"
                self.current += 1
                pass
            pass
        return fresh


class Syntax(abc.ABC):
    """An immutable type that represents the syntax of the DSL"""

    cn = __qualname__

    def __init__(self):
        self.cn = self.__class__.__qualname__
        pass

    def __repr__(self) -> str:
        return f"{self.cn}"

    @abc.abstractmethod
    def to_python(self) -> str:
        return ""

    def __str__(self) -> str:
        return self.to_python()

    def __eq__(self, other: "Syntax"):
        # TODO: Find out what type other should have
        # if it returns the same code, it's the same
        return self.to_python() == other.to_python()

    def __hash__(self):
        # as a bare minimum, let's treat anything that returns the same code as the same
        return hash(self.to_python())

    pass


class Import(Syntax):
    # cn = __qualname__
    imported_module: str

    def __init__(self, mod: str, _as: Optional[str] = None):
        super().__init__()
        self.imported_module = mod
        self._as = _as
        pass

    def __repr__(self) -> str:
        return f"{self.cn}({self.imported_module})"

    def usename(self) -> str:
        if self._as:
            return self._as
        return self.imported_module

    def to_python(self) -> str:
        as_clause = ""
        if self._as is not None:
            as_clause = f" as {self._as}"
            pass
        return f"import {self.imported_module}{as_clause}"

    pass


class FromImport(Import):
    imports: list[str]

    def __init__(self, mod: str, *imports):
        super().__init__(mod)
        self.imports = imports
        pass

    def to_python(self) -> str:
        return f"from {self.imported_module} import {", ".join(self.imports)}"


class Expr(Syntax):
    cn = __qualname__
    pass


primitive = Union[int, float, str]


class Literal(Expr):
    # cn = __qualname__
    def __init__(self, const: Union[primitive, list[Expr]]):
        super().__init__()
        self.const = const
        # if isinstance(const, primitive):
        #     print("primitive")
        #     pass
        # else:
        #     print("list")
        #     pass
        pass

    def is_int(self) -> TypeGuard[int]:
        return isinstance(self.const, int)

    def is_str(self) -> TypeGuard[str]:
        return isinstance(self.const, str)

    def is_float(self) -> TypeGuard[float]:
        return isinstance(self.const, float)

    def is_list(self) -> TypeGuard[list]:
        return isinstance(self.const, List)

    def __repr__(self) -> str:
        return f"{self.cn}({self.to_python()})"

    def to_python(self) -> str:
        if self.is_str():
            return '"' + self.const + '"'
        if self.is_list():
            print(self.const)
            return f"[{", ".join(list(map(lambda x: x.to_python(), self.const)))}]"
        return str(self.const)

    pass


class Variable(Expr):
    # cn = __qualname__
    def __init__(self, name: str):
        super().__init__()
        self.name = name
        pass

    def __repr__(self) -> str:
        return f"{self.cn}({self.name})"

    def to_python(self) -> str:
        return self.name

    def __str__(self) -> str:
        return self.to_python()


class ObjectInit(Expr):
    cn = __qualname__

    def __init__(self, className: str, *args: Expr):
        super().__init__()
        self.className = className
        self.args = args
        pass

    def to_python(self) -> str:
        args = ", ".join(list(map(lambda x: x.to_python(), self.args)))
        return f"{self.className}({args})"

    pass


class AttrAccess(Expr):
    def __init__(self, obj: Expr, key: Expr):
        super().__init__()
        self.obj = obj
        self.key = key
        pass

    def to_python(self) -> str:
        return f"{self.obj}[{self.key}]"

    pass


class ModuleRef(Expr):
    def __init__(self, module: str):
        super().__init__()
        self.module = module
        pass

    def to_python(self) -> str:
        return self.module


class ModuleAccess(Expr):
    def __init__(self, module: Expr, attr: str):
        super().__init__()
        self.module = module
        self.attr = attr
        pass

    def to_python(self) -> str:
        return f"{self.module.to_python()}.{self.attr}"


class ObjectAccess(Expr):
    def __init__(self, e: Expr, prop: str):
        super().__init__()
        self.e = e
        self.prop = prop
        pass

    def to_python(self) -> str:
        return f"{self.e}.{self.prop}"

    pass


class ObjectMethodCall(Expr):
    def __init__(self, v: Variable, method: str, *args: Expr):
        super().__init__()
        self.v = v
        self.method = method
        self.args = args
        pass

    def to_python(self) -> str:
        args = ", ".join(list(map(lambda x: x.to_python(), self.args)))
        return f"{self.v}.{self.method}({args})"

    pass


class KeywordArg(Expr):
    def __init__(self, key: str, e: Expr):
        super().__init__()
        self.key = key
        self.e = e
        pass

    def to_python(self) -> str:
        return f"{self.key}={self.e.to_python()}"


class Statement(Syntax):
    pass

File: react-app/src/dsl/syntax/BasicAst.ts
import * as StringUtils from '../../util/StringUtils';

// Define types
type Primitive = number | string | boolean | null | undefined;
// type Expr = Syntax;
type LiteralType = number | string | boolean | Expr[];

// FreshVar class for generating unique variable names
class FreshVar {
  private current: number;
  private baseName: string;
  private namesToAvoid: Set<string>;

  constructor(baseName: string = "x", namesToAvoid: Set<string> = new Set()) {
    this.current = 0;
    this.baseName = baseName;
    this.namesToAvoid = namesToAvoid;
  }

  getFresh(): string {
    let fresh = `${this.baseName}${this.current}`;
    this.current += 1;
    while (this.namesToAvoid.has(fresh)) {
      fresh = `${this.baseName}${this.current}`;
      this.current += 1;
    }
    return fresh;
  }
}

// Base class for all syntactic elements in the DSL
abstract class Syntax {
  // class name for reference in the __repr__ method
  cn: string;

  constructor() {
    this.cn = this.constructor.name;
  }

  abstract toPython(): string;

  toString(): string {
    return this.toPython();
  }

  // Equality check based on Python representation
  equals(other: Syntax): boolean {
    return this.toPython() === other.toPython();
  }

  // Hash code based on the Python representation
  hashCode(): number {
    return this.toPython().hashCode();
  }
}

abstract class Expr extends Syntax {

}

// Import statement class
class Import extends Syntax {
  importedModule: string;
  private _as: string | null;

  constructor(mod: string, _as: string | null = null) {
    super();
    this.importedModule = mod;
    this._as = _as;
  }

  toPython(): string {
    return `import ${this.importedModule}${this._as ? ` as ${this._as}` : ""}`;
  }

  usename(): string {
    return this._as || this.importedModule;
  }
}

// FromImport statement class
class FromImport extends Import {
  imports: string[];

  constructor(mod: string, ...imports: string[]) {
    super(mod);
    this.imports = imports;
  }

  toPython(): string {
    return `from ${this.importedModule} import ${this.imports.join(", ")}`;
  }
}

// Literal expression class
class Literal extends Syntax {
  const: LiteralType;

  constructor(constant: LiteralType) {
    super();
    this.const = constant;
  }

  isInt(): this is {const: number} {
    return typeof this.const === "number" && Number.isInteger(this.const);
  }

  isStr(): this is {const: string} {
    return typeof this.const === "string";
  }

  isFloat(): this is {const: number} {
    return typeof this.const === "number" && !Number.isInteger(this.const);
  }

  isList(): this is {const: Expr[]} {
    return Array.isArray(this.const);
  }

  toPython(): string {
    if (this.isStr()) {
      return `"${this.const}"`;
    }
    if (this.isList()) {
      return `[${this.const.map((e) => e.toPython()).join(", ")}]`;
    }
    return String(this.const);
  }
}

// Variable expression class
class Variable extends Syntax {
  name: string;

  constructor(name: string) {
    super();
    this.name = name;
  }

  toPython(): string {
    return this.name;
  }
}

// Object initialization expression class
class ObjectInit extends Syntax {
  className: string;
  args: Expr[];

  constructor(className: string, ...args: Expr[]) {
    super();
    this.className = className;
    this.args = args;
  }

  toPython(): string {
    return `${this.className}(${this.args.map((e) => e.toPython()).join(", ")})`;
  }
}

// Attribute access expression class
class AttrAccess extends Syntax {
  obj: Expr;
  key: Expr;

  constructor(obj: Expr, key: Expr) {
    super();
    this.obj = obj;
    this.key = key;
  }

  toPython(): string {
    return `${this.obj.toPython()}[${this.key.toPython()}]`;
  }
}

// Module reference expression class
class ModuleRef extends Syntax {
  module: string;

  constructor(module: string) {
    super();
    this.module = module;
  }

  toPython(): string {
    return this.module;
  }
}

// Module access expression class
class ModuleAccess extends Syntax {
  module: Expr;
  attr: string;

  constructor(module: Expr, attr: string) {
    super();
    this.module = module;
    this.attr = attr;
  }

  toPython(): string {
    return `${this.module.toPython()}.${this.attr}`;
  }
}

// Object access expression class
class ObjectAccess extends Syntax {
  e: Expr;
  prop: string;

  constructor(e: Expr, prop: string) {
    super();
    this.e = e;
    this.prop = prop;
  }

  toPython(): string {
    return `${this.e.toPython()}.${this.prop}`;
  }
}

// Object method call expression class
class ObjectMethodCall extends Syntax {
  v: Variable;
  method: string;
  args: Expr[];

  constructor(v: Variable, method: string, ...args: Expr[]) {
    super();
    this.v = v;
    this.method = method;
    this.args = args;
  }

  toPython(): string {
    return `${this.v.toPython()}.${this.method}(${this.args.map((e) => e.toPython()).join(", ")})`;
  }
}

// Keyword argument expression class
class KeywordArg extends Syntax {
  key: string;
  e: Expr;

  constructor(key: string, e: Expr) {
    super();
    this.key = key;
    this.e = e;
  }

  toPython(): string {
    return `${this.key}=${this.e.toPython()}`;
  }
}

// Base class for all statements
abstract class Statement extends Syntax {}


export type {Primitive, LiteralType}
export {FreshVar, Syntax, Expr, Import, FromImport, Literal, Variable, ObjectInit, AttrAccess, ModuleRef, ModuleAccess, ObjectAccess, ObjectMethodCall, KeywordArg, Statement}